components/server/src/user/token-service.ts

```diff
@@ -42,6 +55,54 @@ export class TokenService implements TokenProvider {
         return promise;
     }
 
+    // EXPERIMENT(sync_refresh_token_exchange)
+    private async doGetTokenForHostSync(userId: string, host: string): Promise<Token | undefined> {
+        const user = await this.userDB.findUserById(userId);
+        if (!user) {
+            throw new ApplicationError(ErrorCodes.NOT_FOUND, `User (${userId}) not found.`);
+        }
+        const identity = this.getIdentityForHost(user, host);
+
+        const doRefreshToken = async () => {
+            // Check: Current token so we can actually refresh?
+            const token = await this.userDB.findTokenForIdentity(identity);
+            if (!token) {
+                return undefined;
+            }
+
+            const aboutToExpireTime = new Date();
+            aboutToExpireTime.setTime(aboutToExpireTime.getTime() + 5 * 60 * 1000);
+            if (!token.expiryDate || token.expiryDate >= aboutToExpireTime.toISOString()) {
+                return token;
+            }
+
+            // Can we refresh these kind of tokens?
+            const { authProvider } = this.hostContextProvider.get(host)!;
+            if (!authProvider.refreshToken) {
+                return undefined;
+            }
+
+            await authProvider.refreshToken(user);
+            return await this.userDB.findTokenForIdentity(identity);
+        };
+
+        try {
+            const refreshedToken = await this.redisMutex.using(
+                [`token-refresh-${host}-${userId}`],
+                2000, // After 2s without extension the lock is released
+                doRefreshToken,
+                { retryCount: 10, retryDelay: 500 }, // We wait at most 10s until we give up, and conclude that we can't refresh the token now.
+            );
+            return refreshedToken;
+        } catch (err) {
+            if (RedisMutex.isLockedError(err)) {
+                log.error({ userId }, `Failed to refresh token (timeout waiting on lock)`, err, { host });
+                throw new Error(`Failed to refresh token (timeout waiting on lock)`);
```

Comment:
Use ApplicationError?

Reply:
I was not sure where it gets handled, so I kept is as before (no ApplkicationError), just changed message.